### PR TITLE
Fixes #69 - NOP Octopus Push from refs/pull/merge

### LINKED ISSUE #69: Don't push packages for `refs/pull/*/merge` branches
Since there are no conditional build tasks in VSTS yet, it would be great to have a check in the Octopus-Push task that will effectively noop if it detects that it is being executed on a build from a `refs/pull/*/merge` branch. This will allow people to reuse a single build definition for CI builds and have their pull request branch policy for `master` (or whatever) set to use this same CI build def.

### PR DESCRIPTION
**Pull request by nshenoy**
Adds a checkbox to the Octopus Push task to allow people to ignore the
execution of the step if the build originates from a pull request merge
branch. This will allow people to use a single build definition for
continuous integration and not worry about packages being pushed.

### CODE REVIEW

File: source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1 [hunk 1]
```diff
@@ -7,12 +7,19 @@ try {
...
     . .\Octopus-VSTS.ps1
 
+    $IgnorePullRequestMergeBuilds = Get-VstsInput -Name IgnorePullRequestMergeBuilds -AsBool
     $OctoConnectedServiceName = Get-VstsInput -Name OctoConnectedServiceName
     $ConnectedServiceName = Get-VstsInput -Name ConnectedServiceName
     $Packages = Get-VstsInput -Name Package -Require
     $AdditionalArguments = Get-VstsInput -Name AdditionalArguments
     $Replace = Get-VstsInput -Name Replace -AsBool
 
+    if( ${env:BUILD_SOURCEBRANCH}  -match "refs/pull/.*/merge$" -and [System.Convert]::ToBoolean($IgnorePullRequestMergeBuilds))
```
Comment (aarondandy) [inline source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1]: Could use a `^` at the start of this regex too.

Comment (aarondandy) [inline source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1]: nit: extra space before `-match`

Comment (aarondandy) [inline source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1]: Could be optimal to check the bool before the regex, not a big deal though.

Comment (aarondandy) [inline source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1]: Will `$IgnorePullRequestMergeBuilds` already be a boolean due to the `-AsBool` above?

File: source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1 [hunk 2]
```diff
@@ -7,12 +7,19 @@ try {
...
     . .\Octopus-VSTS.ps1
 
+    $IgnorePullRequestMergeBuilds = Get-VstsInput -Name IgnorePullRequestMergeBuilds -AsBool
     $OctoConnectedServiceName = Get-VstsInput -Name OctoConnectedServiceName
     $ConnectedServiceName = Get-VstsInput -Name ConnectedServiceName
     $Packages = Get-VstsInput -Name Package -Require
     $AdditionalArguments = Get-VstsInput -Name AdditionalArguments
     $Replace = Get-VstsInput -Name Replace -AsBool
 
+    if( $IgnorePullRequestMergeBuilds -and ${env:BUILD_SOURCEBRANCH} -match "^refs/pull/.*/merge$")
```
Comment (jsquire) [inline source/VSTSExtensions/OctopusBuildAndReleaseTasks/Tasks/Push/Octopus-Push.ps1]: nit:  I think that you may have intended the space to be between the `if` and parenthesis for consistency with surrounding code.

**Comment by Damovisa** [comment]
Hi all,

This change seems really specific to your use case, so I'm reluctant to push it in as is.

Ideally something like this would be handled by the build engine - i.e. skip this task for this branch. The only way to do that now is to have two separate builds filter the branch at the build definition level (as per the below). Either way, I don't think it really belongs in the task itself.

image

That said, to solve the problem, I'm open to it as a more generic advanced option. Like a field where you can supply the regex and we skip the step if the branch matches that regex. Would that suffice?

**Comment by aarondandy** [comment]
Microsoft has not mentioned when that feature will be available. It can be voted on here though: https://visualstudio.uservoice.com/forums/330519-team-services/suggestions/8623600-allow-build-vnext-preview-tasks-to-be-conditionall?tracking_code=69964694473a2c4227461cc80fce6038

**Comment by mud5150** [comment]
Yeah I upvoted this on the VS user voice as well. We have this same need but I agree it belongs in the build engine.

**Comment by Damovisa** [comment]
As we're all agreed it should be a build concern (and that should be possible once Microsoft makes some changes), any objections to closing this PR?

**Pull request closed without merging**